### ISSUE TITLE
Adapt GROUP_CONCAT in Indexation and Reservation mechanism

### ISSUE THREAD

**Issue by naydav** [magento/inventory#158]
Now we using `GROUP_CONCAT` MySQL function in MSI indexation
`\Magento\Inventory\Indexer\StockItem\GetPartialReindexData`

And in our reservation mechanism
`\Magento\Inventory\Model\ResourceModel\Reservation\ReservationCleanup`

But according to documentation
https://dev.mysql.com/doc/refman/5.7/en/group-by-functions.html#function_group-concat

> The result is truncated to the maximum length that is given by the group_concat_max_len system variable, which has a default value of 1024.
But it could be not enough for our purposes

So need to configure MySQL for possibility to use more length in our queries

Solution could be set session variable

> SET [GLOBAL | SESSION] group_concat_max_len = val;

**Comment by toddbc** [comment]
What does this limit mean?  #191 set the limit to 2000, and it looks like this may limit the SKUs associated with a particular "stock".

If I imagine a store with 5000 SKUs in primary stock (please tell me if I'm misunderstanding the relationships here), that probably wouldn't fit in 2000 bytes right?

Just wondering if this imposes any catalog size limits.  I think it'd be unexpected if an administrator added their 286th product, and suddenly stuff stopped working reliably (calculated as 2000 / 7.)

Perhaps we can have the query not using a GROUP_CONCAT, and simply do something like this pseudo-code:
```php
$skusByStock = [];
foreach ($items as $item) {
    $skuList = &$skusByStock[$item[StockSourceLink::STOCK_ID]];
    if (!isset($skuList)) {
        $skuList = [];
    }

    $skuList[] = $item[SourceItemInterface::SKU];
}
unset($skuList);
```

Then, `$skusByStock` could be passed into the foreach in `getStockIdToSkuList()`, which could take the array already built for it.

I don't think this would have any significant performance disadvantage - not one that GROUP_CONCAT wouldn't share.  And it would in theory not have any limitations on product catalog size.  Additionally, since there'd be no string manipulation, it might even use slightly less peak memory.

**Comment by chrom** [comment]
@toddbc it's just test, we will redesign the logic and calculate the necessary limit dynamically

**Comment by franckgarnier21** [comment]
Hello,

This limitation should be highlight somewhere in the documentation.

Indeed, I have configurable product with mode than 100 childs.

So the inventory configurable product source item indexer which retrieve the list of child skus thanks to the GROUP_CONCAT truncates this list. 
Reference: `vendor/magento/module-inventory-configurable-product-indexer/Indexer/SourceItem/SiblingSkuListInStockProvider.php`

This leads to incorrect "quantity, is_salable" value indexed in the inventory_stock_X table for the configurable product.

It is hard to debug and find the issue because the list retrieved by the group_concat has random sort, so in my specific case, the error affects random products at each stock import.

I fixed the issue by increasing the groupConcatMaxLen with the MariaDb default one

`    <type name="Magento\InventoryConfigurableProductIndexer\Indexer\SourceItem\SiblingSkuListInStockProvider">
        <arguments>
            <argument name="groupConcatMaxLen" xsi:type="number">1048576</argument>
        </arguments>
    </type>`

I hope this will help some people to find their issue.

Do you have some update of the new architecture for this point ?

Thanks,

**Comment by sidolov** [comment]
Issue has been resolved in the following commits: fb1b1a513860c7e64a4e6181b1c273e73daa2c1c,
f6078023b7b283fe815d184451299179ff005d82. 
I'm closing it as not relevant anymore.
